### PR TITLE
dx: Render all gridlines (and, bonus: the box cursor) properly

### LINKED ISSUE #543: Terminal: ANSI Underlines are rendered in the wrong color/not at all
* **Your Windows build number:  (Type `ver` at a Windows Command Prompt)**
`Microsoft Windows [Version 10.0.18885.1001]`

* **What you're doing and what's happening: (Copy & paste specific commands and their output, or include screen shots)**
When running `colortest-16` in WSL from inside Windows Terminal, underline sequences are rendered in strange colors (if they get rendered at all). Here is a comparison between the Terminal and the classic console:
Screenshot

* **What's wrong / what should be happening instead:**
The Terminal should render ANSI underlines in the current foreground color.

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
Since we're rendering with antialiasing enabled, we need to make sure
we're stroking actual pixels; to do that, we need to adjust all of our
coordinates by the StrokeWidth / 2. We're always using a stroke width of
1, so that means 0.5.

While I was here, I took the opportunity to fix the color of the grid
lines. Fixes #543.

## PR Checklist
* [x] Closes #543

## Validation Steps Performed

I wrote a quick tool that uses all the gridlines.

image

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
image

**Comment by mdtauk** [comment]
Does the 0.5 number change along with the DPI increasing?

**Comment by DHowett-MSFT** [comment]
@mdtauk no.

**Comment by mdtauk** [comment]
> @mdtauk no.

Intentional, so 1px is 1px even at 200% DPI?

**Comment by DHowett-MSFT** [comment]
@mdtauk One _device-independent pixel unit_ is one dip at any DPI, because it is device-independent. The render target uses standard unscaled coordinates (in DIPs) and the scaling only happens at the time of rasterization.

**Comment by mdtauk** [comment]
> @mdtauk One _device-independent pixel unit_ is one dip at any DPI, because it is device-independent. The render target uses standard unscaled coordinates (in DIPs) and the scaling only happens at the time of rasterization.

Ah so the Direct X pipeline handles the DPI scaling automatically then.  I know XAML uses epx units which adjusts to DPI - I guess I thought coding at that lower level, may need to handle it yourselves.

File: src/renderer/dx/DxRenderer.cpp [hunk 1]
```diff
@@ -295,6 +295,17 @@ DxEngine::~DxEngine()
     RETURN_IF_FAILED(_d2dRenderTarget->CreateSolidColorBrush(D2D1::ColorF(D2D1::ColorF::White),
                                                              &_d2dBrushForeground));
 
+    D2D1_STROKE_STYLE_PROPERTIES strokeStyleProperties{
```
Comment (miniksa) [inline src/renderer/dx/DxRenderer.cpp]:
```suggestion
    const D2D1_STROKE_STYLE_PROPERTIES strokeStyleProperties{
```

File: src/renderer/dx/DxRenderer.cpp [hunk 2]
```diff
@@ -295,6 +295,17 @@ DxEngine::~DxEngine()
     RETURN_IF_FAILED(_d2dRenderTarget->CreateSolidColorBrush(D2D1::ColorF(D2D1::ColorF::White),
                                                              &_d2dBrushForeground));
 
+    D2D1_STROKE_STYLE_PROPERTIES strokeStyleProperties{
+        D2D1_CAP_STYLE_SQUARE,
```
Comment (miniksa) [inline src/renderer/dx/DxRenderer.cpp]: I know I can go look it up, but maybe leave comments on the end of the line explaining what each of the params do? Sometimes we do that for param heavy things.

File: src/renderer/dx/DxRenderer.cpp [hunk 3]
```diff
@@ -1125,6 +1136,13 @@ enum class CursorPaintType
     }
     case CursorPaintType::Outline:
     {
+        // DrawRectangle in antialiasing mode will straddle the pixels; fix that
```
Comment (miniksa) [inline src/renderer/dx/DxRenderer.cpp]: If we get around to doing the issue that lets people choose how much antialiasing they want (cleartype, grayscale, none), will this become a problem that we should watch out for?

Comment (DHowett-MSFT) [reply src/renderer/dx/DxRenderer.cpp]: Fortunately, no! This actually forces the line to come out like it would if AA were off.

Comment (miniksa) [reply src/renderer/dx/DxRenderer.cpp]: Good.

**Pull request merged**